File: Palindrome Linked List/palindromeLinkedList.java
import java.io.*;
import java.lang.*;
import java.util.*;

//This class implements the structure of the Singly linked list
class Node{
	int data;
	Node next;


	Node(int data){
		this.data = data;
		this.next = null;
	}
}


class Main{
	//This function adds two Singly linked list. 
	public static boolean detectPalindromeLinkedList(Node head){
		Stack<Node> stack = new Stack<>();
		Node current = head;
		while(current != null){
			stack.push(current);
			current = current.next;
		}
		current = head;
		while(current != null){
			if (current.data != stack.peek().data) return  false;
			current = current.next;
			stack.pop();
		}
		return true;
	}

	//This function prints the Singly linked list on a single line.
	public static void printList(Node head){
		if(head == null) return ;
		while(head != null){
			System.out.print(head.data + " ");
			head = head.next;
		}

	}

	//This function creates a singly linked list from an array.
	public static Node generateSinglyLinkedList(int[] arr){
		if (arr.length == 0 || arr == null) return null;
		Node SinglyLinkedList = new Node(arr[0]);
		Node temp = SinglyLinkedList;
		for (int i=1;i<arr.length;i++){
			temp.next = new Node(arr[i]);
			temp = temp.next;
		}
		return SinglyLinkedList;
	}
	
	//Main Function
	public static void main(String[] args){
		int[] arr = {1,2,2,1}; //input array
		Node head = generateSinglyLinkedList(arr);
		if (detectPalindromeLinkedList(head)) System.out.println("Given linked list is a palindrome");
		else{
			System.out.println("Given linked list is not a palindrome");
		} 

	}
}

File: 22_Palindrome Linked List/palindromeLinkedList.cpp
#include<bits/stdc++.h>
using namespace std;

// structure of node of linked list
struct Node {

		int data;
		
		Node *next = NULL;
  
};

// Function to check if the linked list
// is palindrome or not
bool isPalindrome(Node* head){
		
		// Temp pointer
		Node* temp= head;

		stack <int> s;

		// Push all elements of the list
		// into the stack
		while(temp != NULL){
				s.push(temp->data);

				// Move ahead
				temp = temp->next;
		}

		// Pop all the elements of stack and
		// Check with elements of linkedlist one by one
		
		while(head != NULL ){
			
			// Get the top most element
			int top =s.top();

			// Pop the element
			s.pop();

			// Check if data is not
			// same as the top element
			if(head -> data != top){
				return false;
			}

		    head=head->next;
		}

    return true;

    
}


int main(){

	// Linked list
	Node* head = new Node;
	head->data = 1;
	Node* temp = head;
  temp->next = new Node;
  temp->next->data = 2;
	temp = temp->next;
	temp->next = new Node;
	temp->next->data = 3;
	temp = temp->next;
	temp->next  = new Node;
	temp->next->data = 2;
	temp = temp->next;
	temp->next  = new Node;
	temp->next->data = 1;
	temp = temp->next;

	if(isPalindrome(head))
			cout<<"Given linked list is a palindrome \n";
	else
		cout<<"Given linked list is not a palindrome\n";

return 0;
}
